Added more options to process_zarr.py.

feng045 opened the pull request:
1. Added these optional inputs to process_zarr.py for more flexibility:
- --s_time: Start datetime to process. Default will process all available times.
- --e_time: End datetime to process. Default will process all available times.
- --z_values: Z values for vertical levels to extract. Default will process all levels.
- --freq: Frequency of the output time steps. Default will process all times.
- --parallel: Flag to run in parallel. Default runs in serial.
- --n_workers: Number of workers to run in parallel.

2. Updated run_process_zarr.sh to include examples of the optional inputs.

feng045 committed:
Added more options to process_zarr.py.

1. Added these optional inputs to process_zarr.py for more flexibility:
- --s_time: Start datetime to process. Default will process all available times.
- --e_time: End datetime to process. Default will process all available times.
- --z_values: Z values for vertical levels to extract. Default will process all levels.
- --freq: Frequency of the output time steps. Default will process all times.
- --parallel: Flag to run in parallel. Default runs in serial.
- --n_workers: Number of workers to run in parallel.

2. Updated run_process_zarr.sh to include examples of the optional inputs.

Commit:
fix: integrate freq input consideration and exact datetime matching in process_zarr.py

- Adjusted logic to ensure that the `start_time` and `end_time` options consider the `freq` input.
  - This ensures accurate extraction of time steps within the specified range, respecting the defined frequency.

- Updated `find_closest_index` function to handle exact matches for datetime values.
  - For datetime values, only exact matches are returned.
  - For numeric types, the function continues to find the closest indices.
  - Added proper conversions and checking to maintain consistency in data types.

laurapaccini commented:
When using the "start_time" and "end_time" options, the script doesn't consider the "frequency" input. 
The new commit needs revision as well, especially for the parallel work.

kosaka90 commented:
Thank you, @feng045 and @laurapaccini, for the improvements. I tried, and it worked great. 
Let's merge the changes to the main. However, we should update a couple of aspects of the output netcdf file:
1) Variable attributes (units and long name) are written in the file (global) attributes; in my original notebook, these attributes are defined in the array **data array**, thus attached to the variable, but in the Python script version, they are defined in the array **dataset**.

2) the time variable as an integer with units of "days" is not helpful. I'm not very familiar with xarray, hence not sure why it converts the numpy datetime64 array with units of seconds (that's what the time variable as read from the Zarr files) into an integer array with units of days while writing to a netcdf file. 

I can fix #1 but not #2...

Also, some dependency information would be valuable in run_process_zarr.sh and readme file. The  run_process_zarr.sh file can load the Python module and activate the pinacles_share environment instead of telling users to prepare an appropriate environment.

feng045 commented:
@kosaka90, both issues you raised on variable attributes and time encoding should be fixed in PR #5.

Specifically,
1. Fixed the bug in output time frequency not matching the input request when `start_time` and `end_time` are specified.
2. Added a new input argument `--ref_time` in process_zarr.py for output netCDF time encoding. Default (not supplied) is ‘2000-01-01T00:00:00’. The time encoding is now ‘seconds since {ref_time}’.
3. Fixed the variable attributes and added a global attribute (time when file was created).

@laurapaccini please review the codes.
